### PR TITLE
Fix #38 issue, update gulpfile.js for gulp v4

### LINKED ISSUE #38: Uncaught (in promise) DOMException: GATT operation already in progress.
Hello there !

I am using your tool for my personal use, it is great and easy to adjust to my needs.

I just want to open an issue for anyone who might stumble upon this unhandled exception as I have found a workaround.

This exception can appear randomly during a firmware update:
`Uncaught (in promise) DOMException: GATT operation already in progress.`
The update stops, and needs to be restarted in order to resume the transfer.

After some investigation, I found out this exception is raised by the characteristic write in the function `sendOperation`.

The workaround I found is the following:

```
private sendOperation(characteristic: BluetoothRemoteGATTCharacteristic, operation: Array<number>, buffer?: ArrayBuffer): Promise<DataView> {
             ...
            characteristic.writeValue(value)
            .catch(() =>  {
                this.log("DOMException: GATT operation already in progress.")
                return Promise.resolve()
                    .then(() => this.delayPromise(500))
                    .then(() => { characteristic.writeValue(value);});
            });
```

Now, everyting works fine and the update continues when the exception is catched.

dfu-update

### PR DESCRIPTION
**Pull request by tkomde**
This PR reflects the proposed code in #38 issue, and corrects the gulpfile.js for gulp V4.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,5 @@ dist
 lib
 types
 npm-debug.log
+package-lock.json
```
Comment (thegecko) [inline .gitignore:8]: Please keep this in the project

File: .gitignore [hunk 2]
```diff
@@ -5,3 +5,5 @@ dist
...
+yarn.lock
```
Comment (thegecko) [inline .gitignore:9]: This project doesn't use yarn, please remove

File: gulpfile.js [hunk 3]
```diff
@@ -1,59 +1,61 @@
...
-var del         = require("del");
-var merge       = require("merge2");
-var buffer      = require("vinyl-buffer");
-var source      = require("vinyl-source-stream");
-var gulp        = require("gulp");
-var sourcemaps  = require("gulp-sourcemaps");
-var typescript  = require("gulp-typescript");
-var tslint      = require("gulp-tslint");
-var uglify      = require("gulp-uglify");
+const path        = require("path");
```
Comment (thegecko) [inline gulpfile.js:1]: :+1:

File: gulpfile.js [hunk 4]
```diff
@@ -1,59 +1,61 @@
...
-function handleError(error) {
+const handleError = error => {
     console.log(error.message);
     if (watching) this.emit("end");
     else process.exit(1);
 }
 
 // Set watching
-gulp.task("setWatch", () => {
+const taskSetWatchFlag = (done) => {
```
Comment (thegecko) [inline gulpfile.js:35]: nit: `(done)` doesn't need the brackets :)

File: package.json [hunk 5]
```diff
@@ -35,20 +35,20 @@
...
+    "gulp-tslint": "^8.1.4",
+    "gulp-typescript": "^5.0.1",
+    "gulp-uglify": "^3.0.2",
     "jszip": "^3.1.3",
     "merge2": "^1.2.0",
     "progress": "^2.0.0",
     "tslint": "5.8.0",
     "tslint-eslint-rules": "^4.1.1",
-    "typescript": "^2.6.2",
+    "typescript": "^3.8.3",
```
Comment (thegecko) [inline package.json:51]: Can you check this still compiles and works for older node versions, it currently supports version 6 and newer I believe

File: src/secure-dfu.ts [hunk 6]
```diff
@@ -294,7 +294,14 @@ export class SecureDfu extends EventDispatcher {
                 reject: reject
             };
 
-            characteristic.writeValue(value);
+            characteristic.writeValue(value)
```
Comment (thegecko) [inline src/secure-dfu.ts:297]: Can you put this in a separate pull request for a clean history?

File: tslint.json [hunk 7]
```diff
@@ -19,7 +19,6 @@
         "no-console": [true],
         "no-empty-interface": false,
         "no-trailing-whitespace": [true],
-        "no-unused-variable": [true],
```
Comment (thegecko) [inline tslint.json:22]: This shouldn't be removed, use underscores in the code if an unused variable appears.

**Review by thegecko — CHANGES_REQUESTED** [review]
Thanks for the pullrequest, @jins-tkomoda 
There's a few minor comments, but mainly we need to check the typescript upgrade still works and this needs splitting into 2 PRs.

Thanks again!

**Comment by tkomde** [comment]
Thank you for your review!
I'll revert this pr and re-submmit separate ones.
